### ISSUE TITLE
stop recording option is not available in scrcpy

### ISSUE THREAD

**Issue by jhachandan1994** [Genymobile/scrcpy#1032]
I am trying to build a script that can automatically get video from scrcpy without opening the scrcpy application is there any way i can do this. is there any way so that scrapy can record video for some specific time lets say 5 - 10 seconds and then automatically exit the program. or atleast any command that can close the recording without manual intervention

**Comment by rom1v** [comment]
There is currently no option to stop automatically (recording or mirroring).

As a workaround, in bash (typically on linux), you could do that with a small script:

```bash
#!/bin/bash
scrcpy -Nr file.mkv &
pid=$!
sleep 10
kill $pid
```

**Comment by jhachandan1994** [comment]
Thanks for the response. I am looking to do this in windows is there any way to do this in windows.

**Comment by rom1v** [comment]
I don't know, probably not.

**Comment by rom1v** [comment]
(ah, yes, open it in MINGW and use a bash script, maybe?)

**Comment by jhachandan1994** [comment]
okay, thanks buddy

**Comment by LBensman** [comment]
You can use `pskill` from [SysInternals](https://docs.microsoft.com/en-us/sysinternals/downloads/pskill) to perform forced termination of the process, similar to how kill works in bash.

**Comment by rom1v** [comment]
But if you kill the process (or if you close the cmd console), the program has no change to clean up (in this case, to finish the recording and write the trailer), contrary to what happens on Linux.

**Comment by jhachandan1994** [comment]
yeah i already tried that taskkill in windows but i am not getting proper results. i was just wondering if i can change the source file to close scrcpy after some interval and then build scrcpy for windows. can you. like we are manually closing scrcpy window so instead i can make scrcpy to close automatically after running for a particular time.

**Comment by rom1v** [comment]
> i was just wondering if i can change the source file to close scrcpy after some interval

Yes. You could add a command-line option `--stop-after` (for example).

You could register a timer ([`SDL_AddTimer`](https://wiki.libsdl.org/SDL_AddTimer)) and send a custom event to the event loop ([here](https://github.com/Genymobile/scrcpy/blob/31bd95022bc525be42ca273d59a3211d964d278b/app/src/scrcpy.c#L130)), where you could do almost the same as `SQL_QUIT`.

**Comment by jhachandan1994** [comment]
Hi thanks for the reply. can you explain further where can i register the timer do i need to modify only scrcpy.c file or i need to change some other files as well.

**Comment by LBensman** [comment]
Oh, is that b/c on Linux build, client installs some sort of signal handler that it lets it finish and write out properly all the buffers?  Sorry, I didn't look at that code closely before recommending.  Also, not sure what format scrcpy writes for captures, but if it is H.264 like what its server sends, then technically you should be able to kill mid stream without corrupting anything written thus far - but yeah, that's not very noble solution :).

B/w another way of doing it without code changes would be to kill scrcpy's server process on the device, assuming client would properly see socket close and terminate nicely.  Process kill on the device can be done with adb without any modification or cooperation from scrcpy.  Just throwing out ideas - no warranty on their goodness ;).

**Comment by jhachandan1994** [comment]
thanks for the idea Lbensman. but i am unable to find any such process running on the device to kill that.

**Comment by jhachandan1994** [comment]
rom1v i have added SDL_AddTimer and hardcoded the same for 5 seconds timer it is working properly in linux and closing after 5 seconds but when bult using makefile it is not working on windows. do i need to do any additional changes for windows built

**Comment by rom1v** [comment]
> B/w another way of doing it without code changes would be to kill scrcpy's server process on the device

Oh, that's true! https://github.com/Genymobile/scrcpy/issues/818#issuecomment-535931801

**Comment by rom1v** [comment]
> rom1v i have added SDL_AddTimer and hardcoded the same for 5 seconds timer it is working properly in linux and closing after 5 seconds but when bult using makefile it is not working on windows. do i need to do any additional changes for windows built

Show the diff.

**Comment by LBensman** [comment]
> thanks for the idea Lbensman. but i am unable to find any such process running on the device to kill that.

As @rom1v linked, the name of process is `app_process` and link will work if it's the only one in the process list (if not, it'll kill other, unrelated, processes started with the same method).  You can get a bit fancier and do some combination of `ps`, `grep` on CLI arguments for scrcpy server to get PID, and then `kill` by PID, rather than `pkill`, for a more targeted and safer operation.

**Comment by jhachandan1994** [comment]
https://github.com/jhachandan1994/scrcpy/blob/83504c0107787eab6502a5d0f278668836031269/app/src/main.c#L32

**Comment by rom1v** [comment]
You should do that after SDL initialization:

https://github.com/Genymobile/scrcpy/blob/31bd95022bc525be42ca273d59a3211d964d278b/app/src/scrcpy.c#L52

**Comment by Maciejszuchta** [comment]
Hello,
Is there a possibility to stop recording the screen on windows? Currently, I have to unplug the device.

**Comment by rom1v** [comment]
@Maciejszuchta Or close the window.

(Unfortunately, on Windows, Ctrl+C just kills the process, and the recording is corrupted.)

**Comment by Maciejszuchta** [comment]
Closing the window also produces a corrupted file. Is there a plan on fixing this issue? Unplugging the device restrains me from using this wonderful library :(

**Comment by rom1v** [comment]
Oh, that's not expected at all. (The scrcpy window, not the console window.)

Which version of scrcpy?

Is there any error message in the console when you close the scrcpy window?

**Comment by rom1v** [comment]
https://github.com/Genymobile/scrcpy/issues/818#issuecomment-535931801

**Comment by Maciejszuchta** [comment]
scrcpy 1.12.1
dependencies:
 - SDL 2.0.10
 - libavcodec 58.54.100
 - libavformat 58.29.100
 - libavutil 56.31.100

**Comment by Maciejszuchta** [comment]
I open 'cmd' window in scrcpy folder and execute the commands from cmd

**Comment by rom1v** [comment]
What command do you execute exactly?

I just tested with v1.12.1. I executed:

```
scrcpy -r file.mkv
```

and

```
scrcpy -r file.mp4
```

I closed the scrcpy window (not the console window), the resulting file is correct in both cases.

**Comment by Maciejszuchta** [comment]
`scrcpy -Nr file.mp4` - since i dont want to display srcpy window

**Comment by rom1v** [comment]
Ah ok, that's expected then :)

See https://github.com/Genymobile/scrcpy/issues/818#issuecomment-535931801 for a workaround.

**Comment by Maciejszuchta** [comment]
Another question is 

> I closed the scrcpy window (not the console window), the resulting file is correct in both cases.

How do you run the commands? Do you open cmd and run `srccpy` ? Or you do it differently?

**Comment by Maciejszuchta** [comment]
> See #818 (comment) for a workaround.

What is app_process ??

**Comment by Maciejszuchta** [comment]
Ok I see. I had to find the `pid` of app_process with `adb shell ps`. And then when I execute
`adb shell pkill {pidofapp_process}` the video is successfully saved. 
Kinda complicated but works :)

**Comment by rom1v** [comment]
If your device has `pkill`, `adb shell pkill app_process` is sufficient (no need to find the actual pid, `pkill` will do it for you).

**Comment by Maciejszuchta** [comment]
Btw is there a possibility to choose the path where the video file is saved? Or I have to move it manually ?

**Comment by rom1v** [comment]
Path the fullpath as parameter instead of just the filename.

**Comment by xzhou29** [comment]
@rom1v 
Is it issue in this file (line#33 - line#35)
scrcpy-master\app\src\sys\unix\file.c
```
memcpy(fullpath, dir, dir_len);
fullpath[dir_len] = '/';
memcpy(fullpath + dir_len + 1, file, file_len + 1);
```

**Comment by shaikhabdulkarim** [comment]
You can stop the recording by turning off the usb tethering option in your Android device.

**Comment by evillurker** [comment]
For context if anyone stumbles across this and has windows, you have to taskkill /f /im adb.exe and then it will properly finish. I thought I had to type adb shell pkill which didn't work.
